Feature/master post filters 9184

onepointsixtwo opened the pull request:
Fixes #9184 

This change adds the ability to view the list of blog posts in a compact view with a toggle in the top right hand side of the screen, to move between the new compact view and the standard view which is already present in the application. The new view has only a title, the date of the article and a list of statuses shown, as well as a much smaller image. All of the actions are hidden behind a menu using a vertical ellipsis on the right hand side of the row. The view type is persisted so that on returning to the view it remains the same.

To test:

1. Open the app and open blog posts
2. Make sure there are visible blog posts within the current section
3. Hit the button in the top right and check that the view displays as the compacted version (check against the design in the issue #9184) 
4. Hit the button again and make sure the view displays as the original version again
5. Hit the button one more time to return the view to compact
6. Kill the app (swipe out of multi-tasking)
7. Return to the app and make sure the view still displays as compact
8. Open the menu on the right hand side of the compact view, and make sure all button options are present, as they would be on the non-compacted versions.

Update release notes:

- [Y] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

onepointsixtwo commented:
![posts-toggle-demo](https://user-images.githubusercontent.com/5874664/56310836-b5f65180-6144-11e9-8cd7-16dc7bb467d8.gif)

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt:
```diff
@@ -11,30 +11,37 @@ import android.view.ViewGroup
     private val postViewHolderConfig: PostViewHolderConfig,
-    private val uiHelpers: UiHelpers
+    private val uiHelpers: UiHelpers,
+    private var viewLayoutType: ViewLayoutType
```
This approach works as expected, however, I believe the `PostListItemUiState` doesn't match the UI now. 

For example when the list item has 5 actions available the `actions` field contains the following structure:
```
(SingleItem,SingleItem, MoreItem(SingleItem,SingleItem,SingleItem))
```
The UI presents it as
```  
(MoreItem(SingleItem,SingleItem,SingleItem,SingleItem,SingleItem))
```

What do you think about introducing new `PostListItemType` for the compact view. We wouldn't need to pass the `ViewLayoutType` in the constructor and the view model (more specifically `PostListItemUiStateHelper`) would take care of creating the correct state. I believe it'd be more consistent with the other view holder types `LoadingItem and EndListIndicatorItem`. 

Having said that, maybe it'd be too complicate and it's not worth it. However, I think we should at least try to make the `PostListItemUiState` match the UI. Wdyt? Do you have any better ideas how to go about it?

onepointsixtwo replied:
That is true. I think you're right really - there want to be different view objects depending on the layout. It felt when I wrote it that shouldn't really be the case, as effectively we are displaying the same information but in a different shape, so really it's entirely a view decision and not a view model (as in theory, the model hasn't changed) but you're right that the menu broke that at the last minute.
I guess there are two options: 
1) Use a totally different model for the compact rows or 
2) Include a key in there for compactActions as well as actions. 
I actually don't mind doing either of these things. What's your preference?

malinajirka replied:
I usually prefer to have more structure than less, so if it were up to me I'd go with "1". The two layouts share most of the data, but not all of them (eg. the compact layout doesn't have excerpt). 

I think I'd need to see both version to be sure whether it feels right. So it's up to you, feel free to go with the easier solution if it feels like the right one.

malinajirka commented on build.gradle:
```diff
@@ -9,7 +9,7 @@ buildscript {
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.2.1'
+        classpath 'com.android.tools.build:gradle:3.3.2'
```
I think this change isn't related to the "toggle" feature. It also introduces some lint issues and fixing those within this PR might make it more difficult to review. I believe it should be taken care of in its own PR. We always try to keep our PRs small and single-focused. Wdyt?

onepointsixtwo replied:
True, I'll pull this out ASAP.

malinajirka replied:
👍

malinajirka commented on WordPress/src/main/res/layout/post_list_item_compact.xml:
```diff
@@ -0,0 +1,108 @@
+<?xml version="1.0" encoding="utf-8"?>
+<layout xmlns:android="http://schemas.android.com/apk/res/android"
```
You might want to consider checking the dimens.xml for existing values and reusing them or creating new values so we don't have hardcoded dimensions in layout files. Wdyt?

onepointsixtwo replied:
Sure, I'll resolve this. Good point, and my bad.

malinajirka replied:
👍

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt:
```diff
@@ -11,30 +11,37 @@ import android.view.ViewGroup
+private const val VIEW_TYPE_ENDLIST_INDICATOR = 2
+private const val VIEW_TYPE_LOADING = 3
 
 class PostListAdapter(
```
I'm not sure what you mean by this one; sorry! I think the comment maybe doesn't match up with the line (?)

malinajirka replied:
Sorry, my bad. I should have included a link to the code I'm referring to.


Should we update [this](https://github.com/wordpress-mobile/WordPress-Android/blob/dfe83b8b7236483e3fc12352fd95fb61f2bced32/WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt#L101) class so the items are considered as same items with different contents? Wdyt?

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/ViewLayoutType.kt:
```diff
@@ -0,0 +1,16 @@
+package org.wordpress.android.ui.posts
+
+enum class ViewLayoutType(val id: Long) {
```
Sure, OK. I don't mind general too much in general (no pun intended :D) for Android because it's namespaced anyway as opposed to iOS where the names end up really long, but if you'd prefer it given a name to fit with its namespace that's all good. I don't have a strong preference really.

malinajirka replied:
>I don't mind general too much in general (no pun intended :D)
:D

> if you'd prefer it given a name to fit with its namespace that's all good.

This might be my personal preference, rather than a general practice in our code base. However, I feel the autocompletion can become a bit useless as there are more classes with general names - it's also pretty easy to unintentionally import a wrong class.

I'll leave the final decision to you. I don't feel strongly about it.

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.kt:
```diff
@@ -309,4 +336,17 @@ class PostsListActivity : AppCompatActivity(),
+
+    // Menu ViewLayoutType handling
+
+    private fun updateMenuIconForViewLayoutType(viewLayoutType: ViewLayoutType) {
```
This is the only thing I might disagree with; I think this logic is already as much in the ViewModel as can be, because the actual icon reference should not be in the ViewModel layer at all. The view model should have no knowledge of the inner working of Android systems so that it can still be tested in a pure unit test environment rather than an Android Test environment. Unless I'm missing something, attempting to move this into the view model would required the view model having knowledge about the actual icons?

malinajirka replied:
>The view model should have no knowledge of the inner working of Android systems so that it can still be tested in a pure unit test environment rather than an Android Test environment.

I can't agree more with this sentence.

> attempting to move this into the view model would required the view model having knowledge about the actual icons?

I don't believe that's the case. The id of an icon is just a simple Integer from a generated java file - `R.java`. We use both string and drawable resources (their ids to be more specific) in our view models and unit tests work as expected.

onepointsixtwo replied:
Ah interesting, I never use R files in unit tests since they seemed so Android specific. OK, sure, I'll change this, less untestable logic is never a bad thing!

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListItemCompactViewHolder.kt:
```diff
@@ -0,0 +1,108 @@
+package org.wordpress.android.ui.posts
```
Completely agree, will refactor this to share the shared parts

malinajirka replied:
👍

onepointsixtwo commented on WordPress/build.gradle:
```diff
@@ -117,6 +117,11 @@ android {
         // MPAndroidChart uses androidX - remove this line when we migrate everything to androidX
         exclude 'META-INF/proguard/androidx-annotations.pro'
     }
+
```
OK, so this is obviously the big comment, which is why I'm coming to replying to it last of the lot. I had to think about this a bit.

OK, so you're right that it's a new architectural direction in a sense, but not a prescriptive one. By this I mean the decision to use Android view binding would not force the update of any parts of the application which were not already using it. It simply creates files in the build directory using the preprocessor which are ignored by anything which doesn't use them.
I would therefore really really like to include its usage. As you can see, it drops the number of lines of non-declarative and not easily re-usable code by a large margin. I do completely understand the objection of using mixed arch in the same view, but if this is the _only_ problem then I would actually much rather refactor the existing than let the existing determine what should be used for the new.

_Having said all of the above_; if you're saying that you'd rather I did not attempt to introduce an architectural change at all, then I can write it the old-fashioned way, but being completely honest, I'd rather not. I actually think including view bindings can produce the cleanest arch you can get for Android, and is not a third party library for which you'd have to consider support being dropped etc. If it is something that would be considered realistically, then I'm happy to write the p2 post explaining it and provide a strong case for why I think it's a good idea with code examples.

malinajirka replied:
This discussion is starting to have two parts
1. Should we use data binding in our codebase?

I'm not against it in general. I have a couple of thoughts about it, but I've used it in a large project and it worked pretty well. I'd rather not go into details in this thread as it could easily get out of hand. Having said that, I'll be more than happy to continue with this discussion on a p2 post/public slack channel.

2. Should we introduce data binding in this PR?

This is the part I feel strongly about. I don't think introducing data binding should be our decision, but rather a decision of the whole team (or several members at least). 

>you're right that it's a new architectural direction in a sense, but not a prescriptive one.

Yes, nothing forces you to use it everywhere, but at the same time it brings yet another inconsistency into our code base. Moreover, I believe it increases the build time and has other side effect which I feel should be considered and discussed first.

Let's ask a third person about their thoughts since we apparently don't share the same opinion, which is completely fine. @oguzkocer Would you mind chiming in? Thanks!

oguzkocer replied:
I agree with @malinajirka that we shouldn't just throw in a dependency without giving other developers a chance to comment on it. Whether data-binding is the right way to go or not is not really the point of this discussion, so my opinion on that part doesn't really matter. However, for some context, I know that we tried data-binding a while back and had some issues with it, so we decided to put it off. I imagine those issues would already be addressed, so there is probably a good case to made for it, but I'd prefer this discussion to happen outside of a PR.

@onepointsixtwo If you feel strongly about data binding and more importantly if you don't feel comfortable with proposing this PR without it, please feel free to start a discussion for it while we put this PR on hold. Otherwise, I suggest making the necessary changes to this PR _and then_ starting the discussion for data-binding so we can keep things going.

Outside of the discussion of data-binding, I think it's important that we are all able to work within the constraints and follow patterns even if we don't feel comfortable with them. In a code base like WordPress, I can't stress the importance of that enough.

Hope this helps!

onepointsixtwo commented:
OK, for Monday, I've addressed all the above change requests including:

- Go through _post_list_item_compact.xml_ and use _dimens.xml_ values, not hard coded ones.
- Undo change to Gradle version
- Do not use Android view binding
- Change name of *ViewLayoutType* to something more specific to posts (*PostListViewLayoutType*)
- Use the view model for menu icon logic in the posts activity
- Share code between the view holders to DRY out the code re-use cases
- Update DiffUtil implementation
- Don't use PostListItemUiState for a UI for which it does not fit

Three of these took a little more decisioning than the others:

### PostListItemUiState Does Not Match UI

This is one of the big ones really. This wasn't entirely possible to separate out, because of the way all the helper classes being used work. The cached list of paged things is in the adapter, not in the ViewModel. This means that we only store a cache of the already transformed data, not the original. I tried various things to re-transform the original, but it doesn't seem to be possible without writing some quite big changes to PostListItemDataSource's superclass to start storing a cache of the paged data that can be then re-transformed into either the compact or the non-compact version. I can do this but I held off for now for two reasons:

1) It would take a while
2) I'm not sure if it's actually desirable - the Peril system is already giving a warning regarding too many changes for a single PR so I didn't want to make this change unless I was really certain this was actually wanted.

I went with a simpler technique therefore, of the data object containing information (separately) for both UI versions and the UI using the appropriate version for what was being displayed. The advantage being this requires no new transform of the input data, and can be switched with low overhead. It does however mean (some minimal amount) of logic in the adapter. I'd like to know your thoughts on this?


### Share code between view holders

I've done this but I actually did use inheritance in the end, and added a shared base class for the common parts. It actually didn't seem a bad case for direct inheritance. The relationship seemed like an is-a rather than a has-a in the sense that both of them are-a row representing a blog row in different ways. Let me know if what you think!

### Update DiffUtil implementation

Didn't need updating after the changes but had to add manual equals() / hashCode() to the post list data classes now they are not data classes (because of their requirement to inherit from a shared class). 

### Tests

Updated the only test which contained anything relevant to also test for the compacted version which was postListItemUiStateHelper.kt. Was unable to run the UI tests due to credentials issue so I haven't added any.


I think that's everything - let me know if there's anything I've missed.

oguzkocer commented:
@onepointsixtwo I didn't see your [last comment](https://github.com/wordpress-mobile/WordPress-Android/pull/9657#issuecomment-484943337) before I wrote [mine](https://github.com/wordpress-mobile/WordPress-Android/pull/9657#pullrequestreview-228765715) as I started it a while back but got distracted with some other things. So please consider my comments in that context. cc @malinajirka

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.kt:
```diff
@@ -195,6 +200,7 @@ class PostsListActivity : AppCompatActivity(),
         viewModel.updatePostsPager.observe(this, Observer { authorFilter ->
             authorFilter?.let {
                 val currentItem: Int = pager.currentItem
+                val viewLayoutType = viewModel.viewLayoutType.value ?: PostListViewLayoutType.defaultValue
```
True, must have missed it when refactoring. Removed.

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.kt:
```diff
@@ -287,10 +298,25 @@ class PostsListActivity : AppCompatActivity(),
 
+    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
+        menu?.clear()
+        menu?.add(Menu.FIRST, R.id.post_menu_item_view_layout_type, 3, "")?.let { item ->
```
There is actually a reason I didn't do this; I started out by using an XML file to define the menu, but despite the fact I used showAsAction="always" it never showed - it just hid behind an ellipsis menu (...) which was terrible UI when there was only one button. I honestly couldn't work out why - I've never had problems with showAsAction always before. Is there any global style within the project that you know of that could be causing this?

malinajirka replied:
That's weird. Afaik we don't have any global style or anything similar which might cause these issues.

I have quickly tried to implement it using xml and it worked as expected.

```
menu?.let {
            menuInflater.inflate(R.menu.test_menu, menu)
            toggleViewLayoutMenuItem = menu.findItem(R.id.post_menu_item_view_layout_type)
        }
        return super.onCreateOptionsMenu(menu)
```

However, if it doesn't work for you, don't worry about it and we can improve it later. ;)

malinajirka replied:
See https://github.com/wordpress-mobile/WordPress-Android/pull/9657#pullrequestreview-230638491

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.kt:
```diff
@@ -287,10 +298,25 @@ class PostsListActivity : AppCompatActivity(),
         return super.onOptionsItemSelected(item)
     }
 
+    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
```
Added

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListMainViewModel.kt:
```diff
@@ -152,6 +161,10 @@ class PostListMainViewModel @Inject constructor(
     init {
         lifecycleRegistry.markState(Lifecycle.State.CREATED)
+
+        val layout = prefs.postListViewLayoutType
```
Sure, good point, much cleaner :-)

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListMainViewModel.kt:
```diff
@@ -95,6 +98,12 @@ class PostListMainViewModel @Inject constructor(
+    private val _viewLayoutType = MutableLiveData<PostListViewLayoutType>()
+    val viewLayoutType: LiveData<PostListViewLayoutType> = _viewLayoutType
+
+    private val _updateViewLayoutMenuIcon = MutableLiveData<Int>()
```
Changed

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListViewModelConnector.kt:
```diff
@@ -13,8 +13,8 @@ class PostListViewModelConnector(
     val doesPostHaveUnhandledConflict: (PostModel) -> Boolean,
-    val postFetcher: PostFetcher,
-    private val getFeaturedImageUrl: (site: SiteModel, featuredImageId: Long, postContent: String) -> String?
+    private val getFeaturedImageUrl: (site: SiteModel, featuredImageId: Long, postContent: String) -> String?,
```
No, I think it was because one of the mid-stages of development I added a third parameter. I'll change it back.

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListMainViewModel.kt:
```diff
@@ -330,4 +343,20 @@ class PostListMainViewModel @Inject constructor(
+        _updateViewLayoutMenuIcon.value = iconForViewLayout(toggledValue)
+    }
+
+    private fun iconForViewLayout(layout: PostListViewLayoutType) = when (layout) {
```
I'm not sure I understand this one; so at the moment, the values are drawable resources, but UIDimenRes represents a dimension resource? I could make a data class to ensure that the int was actually a drawable resource (e.g. data class DrawableRes(@DrawableRes id: Int))?

malinajirka replied:
Sorry, my bad. Please ignore this comment.

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt:
```diff
@@ -58,51 +61,83 @@ class PostListItemUiStateHelper @Inject constructor(private val appPrefsWrapper:
-                        postStatus = postStatus,
-                        isLocalDraft = post.isLocalDraft,
-                        isLocallyChanged = post.isLocallyChanged,
+        val actions = createActions(
```
Good point. Fewer lines is always better :-)

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt:
```diff
@@ -20,21 +26,52 @@ import org.wordpress.android.viewmodel.posts.PostListItemType.LoadingItem
+
+    init {
+        layout = postListViewLayoutType.value ?: PostListViewLayoutType.defaultValue
+        postListViewLayoutType.observe(fragment, Observer {
```
Sure, I was in two minds as to which to do anyway. I was trying to avoid two places having to know about it, but I agree that it doesn't fit in with the way the other property observations work. Updated.

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt:
```diff
@@ -20,21 +26,52 @@ import org.wordpress.android.viewmodel.posts.PostListItemType.LoadingItem
+    }
+
+    init {
+        layout = postListViewLayoutType.value ?: PostListViewLayoutType.defaultValue
```
Renamed and moved.

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt:
```diff
@@ -272,7 +308,7 @@ class PostListItemUiStateHelper @Inject constructor(private val appPrefsWrapper:
         }
 
-        return if (buttonTypes.size > MAX_NUMBER_OF_VISIBLE_ACTIONS) {
+        return if (buttonTypes.size > MAX_NUMBER_OF_VISIBLE_ACTIONS && postListViewLayoutType != COMPACT) {
```
Done - it no longer takes an input of the view layout type  - jus the number of visible actions which for compact is set to unlimited (Int.MAX_VALUE)

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemType.kt:
```diff
@@ -10,27 +10,49 @@ import org.wordpress.android.widgets.PostListButtonType
 sealed class PostListItemType {
     class PostListItemUiState(
         val data: PostListItemUiStateData,
+        val compactData: PostListItemCompactUiStateData,
```
OK, made the appropriate changes here. It removes a reasonable number of lines. Also added in the progress / disabled. I actually just missed what they did exactly in the main view. I had to hard code the values to test this - I'm not actually sure how you get them to show in the course of actual events.

malinajirka replied:
In order to show the progress/disabledOverlay you can for example click on the Publish button on an item in the Drafts tab.

onepointsixtwo commented:
Hi @malinajirka - thanks for the continued reviews :-) I'm going to have to finish for this evening, without having quite finished making changes. However, I'm getting there. I had a few queries when you get a chance, where I wasn't quite sure of something or other. 

Also, apologies. I started marking the threads as resolved so I could easily keep track of where I was but I realised after this might make it a little difficult to see them for you - sorry if it does! The ones I haven't marked as resolved are the ones I still have queries on in some way.

I definitely agree with the compact skeleton for loading. I haven't started the second PR for this yet, but I'll try and get to it tomorrow evening. It seems fairly straightforward having looked at the implementation for the non-compact skeleton.

onepointsixtwo commented:
Changed my mind about the compact skeleton work - I decided it was quick enough to finish off before bed. Here's the PR: https://github.com/onepointsixtwo/WordPress-Android/pull/1 though it's currently a PR with the base branch set to the branch for this PR (simply so the changes could be seen from the point of view of differences to this PR rather than the original branch which would include this work too). I can update the base branch later to be a PR against the correct branch.

malinajirka commented:
Thanks for the changes to the PR and for creating a new PR for the skeleton view! I have replied to the open discussions, but I wasn't sure whether it's ready for another review as you mentioned "I'm going to have to finish for this evening, without having quite finished making changes.". I'd rather review it when you make all the changes you are planning to if you are fine with it. ;)

>Also, apologies. I started marking the threads as resolved so I could easily keep track of where I was but I realised after this might make it a little difficult to see them for you - sorry if it does!

No worries;). However, as you mentioned it's easier to review when the reviewer is the one who resolves the discussions. I usually simply add 👍 reaction or comment with "Fixed in commit #abc" to keep track of the issues I fixed.

onepointsixtwo commented:
To be honest, the only things to resolve were the ones I had questions about, so if you don't want me to change the UIDimenRes thing then I think I might be finished. I just tried again using the XML layout for the menu and got exactly the same thing. It's really odd. I used pretty much exactly the same code as you. For interest, here's the XML:

`<?xml version="1.0" encoding="utf-8"?>
<menu xmlns:android="http://schemas.android.com/apk/res/android"
      xmlns:tools="http://schemas.android.com/tools">
    <item
        android:id="@+id/post_menu_item_layout_type"
        android:icon="@drawable/ic_view_post_compact"
        android:title="@string/posts"
        tools:showAsAction="always" >
    </item>
</menu>`

I honestly have no idea why it's not working. 

Other than that, I think I'm ready for re-review when you are :-)

malinajirka commented on WordPress/src/main/res/drawable/ic_view_post_compact.xml:
```diff
@@ -0,0 +1,9 @@
+    android:viewportWidth="24"
+    android:viewportHeight="24">
+  <path
+      android:fillColor="#FF000000"
```
I think we might want to change the color to `@android:color/white` and rename the drawable to `ic_view_post_compact_white_24dp.xml` in order to keep it consistent with other drawables.

onepointsixtwo replied:
Changed.

malinajirka replied:
Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/9657/commits/762018575fe953b599121659208dcf1ba5f80498

malinajirka commented on WordPress/src/main/res/drawable/ic_view_post_full.xml:
```diff
@@ -0,0 +1,9 @@
+    android:viewportWidth="24"
+    android:viewportHeight="24">
+  <path
+      android:fillColor="#FF000000"
```
I think we might want to change the color to `@android:color/white` and rename the drawable to `ic_view_post_full_white_24dp.xml` in order to keep it consistent with other drawables.

onepointsixtwo replied:
Also changed.

malinajirka replied:
Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/9657/commits/762018575fe953b599121659208dcf1ba5f80498

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.kt:
```diff
@@ -309,4 +335,13 @@ class PostsListActivity : AppCompatActivity(),
+
+    private fun updateMenuIconForViewLayoutType(@DrawableRes iconId: Int) {
+        getDrawable(iconId)?.let { drawable ->
+            drawable.setTint(Color.WHITE)
```
I think we can remove this line now as the drawables are already white. If we decide not to remove it we might want to consider passing the tintColor from the view model. Wdyt?

onepointsixtwo replied:
Done. No reason to have it anymore so I'll keep it super simple.

malinajirka replied:
Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/9657/commits/762018575fe953b599121659208dcf1ba5f80498

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt:
```diff
@@ -72,6 +96,11 @@ class PostListAdapter(
     }
 
+    fun updateItemLayoutType(updatedItemLayoutType: PostListViewLayoutType) {
+        itemLayoutType = updatedItemLayoutType
```
We might want to consider checking whether `updatedItemLayoutType` is actually different than `itemLayoutType`. Wdyt?

onepointsixtwo replied:
Sure, no point in updating data if it's the same (though it of course shouldn't really happen since it's a toggle!). Added an early exit just in case though.

malinajirka replied:
Ohh, you are right. I believe it was being called more than once before we changed this https://github.com/wordpress-mobile/WordPress-Android/pull/9657#discussion_r278543502.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt:
```diff
@@ -37,7 +40,8 @@ import org.wordpress.android.widgets.PostListButtonType.BUTTON_SYNC
 
-private const val MAX_NUMBER_OF_VISIBLE_ACTIONS = 3
+private const val MAX_NUMBER_OF_VISIBLE_ACTIONS_STANDARD = 3
+private const val MAX_NUMBER_OF_VISIBLE_ACTIONS_COMPACT = Int.MAX_VALUE
```
This seems a bit confusing, especially since the number of visible actions on the compact view is `1`.
I don't want to re-open [this discussion](https://github.com/wordpress-mobile/WordPress-Android/pull/9657#discussion_r277591246) but I'd like to hear your opinion. What I had in mind is basically create the following structure

Standard (SingleItem,SingleItem,MoreItem(SingleItem,SingleItem,...))
Compact(MoreItem(SingleItem,SingleItem,...))

What do you think about it?

onepointsixtwo replied:
Ah I see what you mean. Involved some slight re-jigging, but that is cleaner. And you're right - the menu is effectively a more menu in compact - it's all hidden behind an ellipsis which is the definition of a more menu in the standard view.

malinajirka commented on WordPress/src/main/res/menu/edit_comment.xml:
```diff
@@ -1,8 +1,10 @@
         android:title="@string/save"
-        app:showAsAction="always" />
+        app:showAsAction="always"
+        tools:ignore="AppCompatResource"/>
```
I'm a bit confused by this change - it seems the warning says we should use "app" namespace instead of "android" when we are using AppCompat. Why would we want to suppress the warning when we already use the correct namespace? What am I missing?

onepointsixtwo replied:
Nothing, I don't know why this is changed either. Wondering if it's something auto-generated by IntelliJ or something - I don't feel like it's something I'd change. Undone anyway.

malinajirka commented on WordPress/src/main/res/layout/post_list_item_compact.xml:
```diff
@@ -0,0 +1,115 @@
+            android:layout_marginStart="@dimen/margin_small"
+            android:layout_marginTop="@dimen/margin_large"
+            android:layout_toStartOf="@+id/more_button"
+            android:background="@color/red_0"
```
Do we intentionally set the background to red color?

onepointsixtwo replied:
Ah - I did for reasons of laying the view out and seeing the preview (since I couldn't see white-on-white) but I'll change it to transparent now I'm done

malinajirka replied:
I think we could just remove the background attribute, but it doesn't really matter. 👍

malinajirka commented on RELEASE-NOTES.txt:
```diff
@@ -4,6 +4,7 @@
 * Redirect to username/password login when WordPress.com reports email login not allowed
 * Updated notifications list to Material guidelines
 * Updated snackbar design to Material guidelines
+* Added ability to toggle to compact view type for blog posts list
```
Note for myself - we need to move this line to 12.4 when we merge develop into this branch.

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt:
```diff
@@ -104,6 +104,13 @@ class PostListFragment : Fragment() {
         val mainViewModel = ViewModelProviders.of(nonNullActivity, viewModelFactory)
                 .get(PostListMainViewModel::class.java)
 
+        postListAdapter.updateItemLayoutType(mainViewModel.viewLayoutType.value ?: PostListViewLayoutType.defaultValue)
```
We don't it seems - I didn't realise the observers fire with initial value but tested after our convo and they definitely do.

malinajirka replied:
Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/9657/commits/762018575fe953b599121659208dcf1ba5f80498

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListItemViewHolder.kt:
```diff
@@ -22,79 +24,110 @@ import org.wordpress.android.viewmodel.posts.PostListItemAction
-                        onMoreClicked(action, view)
-                    }
-                }
+        private var currentItem: PostListItemUiState? = null
```
Good point. Changed to the same way as the standard one works, and stripped out more lines :-)

malinajirka replied:
Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/9657/commits/762018575fe953b599121659208dcf1ba5f80498

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListMainViewModel.kt:
```diff
@@ -157,6 +166,10 @@ class PostListMainViewModel @Inject constructor(
         this.site = site
 
+        val layout = prefs.postListViewLayoutType
+        _viewLayoutType.value = layout
```
Kept it simple and moved into method called SetViewLayoutAndIcon()

malinajirka replied:
Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/9657/commits/762018575fe953b599121659208dcf1ba5f80498

onepointsixtwo commented on WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt:
```diff
@@ -20,21 +23,39 @@ import org.wordpress.android.viewmodel.posts.PostListItemType.LoadingItem
 ) : PagedListAdapter<PostListItemType, ViewHolder>(PostListDiffItemCallback) {
     private val layoutInflater: LayoutInflater = LayoutInflater.from(context)
+    private var itemLayoutType: PostListViewLayoutType = PostListViewLayoutType.defaultValue
+    private var recyclerView: RecyclerView? = null
```
Done, sorry, silly miss.

malinajirka replied:
Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/9657/commits/762018575fe953b599121659208dcf1ba5f80498

onepointsixtwo commented:
Ok, addressed all of the above. All of the smaller issues were really straightforward and I've removed any unnecessary code, and changed the icon names / fill colours, put in a method to set the values for layout atomically in PostListMainViewModel.kt.

I also did the bigger change because you're right that it's cleaner to use the MoreAction type to represent the compact actions. It involved a little re-jigging to allow the method to accept the idea of having 0 actions on display, but it's cleaner and makes more sense.

onepointsixtwo commented:
I forgot to say this but I also tried using the XML layout again for menu. It still hides it behind an ellipsis, even changing the XML namespace to app (res-auto). Here's all the code just in case you can see anything: https://github.com/onepointsixtwo/WordPress-Android/compare/feature/master-post-filters-9184...onepointsixtwo:misc/menu-test?expand=1 
I tried including menu?.clear() and removing it in case there was something down the chain overriding the menu's style. Either way it still hides the menu like in the attached screenshot.
![Screenshot 2019-04-26 at 09 10 14](https://user-images.githubusercontent.com/5874664/56793137-2ebf7280-6803-11e9-9d16-e8b9a35ada2d.png)

malinajirka commented:
Thanks for the changes;)! 

I'm going to merge this PR as we are merging it into a working branch anyway. I've left 2 super minor nitpicks and one rather important comment. I'll create a separate issue for them and we can fix them in another PR.

You might want to consider separating each suggestion into its own commit unless the change is really minor in the future. It makes the changes much easier to review.


**Regarding the issue with the menu**
The problem is you are not using `SupportMenuInflator`. If you combine the code I posted in [here](https://github.com/wordpress-mobile/WordPress-Android/pull/9657#discussion_r277972142)  and the [namespace tip](https://github.com/wordpress-mobile/WordPress-Android/pull/9657#pullrequestreview-230638491) it should work. You could also replace `MenuInflator(this)` with `SupportMenuInflator(this)`, but since the parent activity has getter for the inflator, it doesn't make much sense. 

I also don't think we need the `menu?.clear()` call. We'd basically clear all the menu items the parent has just added (I know the parent currently doesn't add any menu items, but you get my point;)). 

Just a heads up: You'll want to update the icon after the menu is inflated. You might want to consider moving the subscription to the `onCreateOptionsMenu`.


Thank you for implementing the feature and for implementing/discussing all my suggestions 🙇 !

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListItemViewHolder.kt:
```diff
@@ -22,79 +24,99 @@ import org.wordpress.android.viewmodel.posts.PostListItemAction
+
+            itemView.setOnClickListener { item.onSelected.invoke() }
+
+            if (item.compactActions.actions.isNotEmpty()) {
```
Tracked here https://github.com/wordpress-mobile/WordPress-Android/issues/9724

malinajirka commented on WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt:
```diff
@@ -268,19 +286,39 @@ class PostListItemUiStateHelper @Inject constructor(private val appPrefsWrapper:
+        buttonTypes: List<PostListButtonType>,
+        onButtonClicked: (PostListButtonType) -> Unit
+    ): MoreItem {
+        val createSinglePostListItem = { buttonType: PostListButtonType ->
```
Tracked here https://github.com/wordpress-mobile/WordPress-Android/issues/9724

malinajirka commented on WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt:
```diff
@@ -24,6 +24,7 @@ import org.wordpress.android.ui.utils.UiString.UiStringRes
 import org.wordpress.android.ui.utils.UiString.UiStringText
 import org.wordpress.android.util.AppLog
 import org.wordpress.android.util.AppLog.T.POSTS
+import org.wordpress.android.viewmodel.posts.PostListItemAction.MoreItem
```
Tracked here https://github.com/wordpress-mobile/WordPress-Android/issues/9724

malinajirka commented:
I've also submitted a ticket for the options menu - https://github.com/wordpress-mobile/WordPress-Android/issues/9725
